app/discovery/provider/docker.go

```diff
@@ -272,3 +257,79 @@ func contains(e string, s []string) bool {
 	}
 	return false
 }
+
+type dockerClient struct {
+	client  http.Client
+	network string // network for IP selection
+}
+
+// NewDockerClient constructs docker client for given host and network
+func NewDockerClient(host, network string) DockerClient {
+	var schemaRegex = regexp.MustCompile("^(?:([a-z0-9]+)://)?(.*)$")
+	parts := schemaRegex.FindStringSubmatch(host)
+	proto, addr := parts[1], parts[2]
+	log.Printf("[DEBUG] Configuring docker client to talk to %s via %s", addr, proto)
+
+	client := http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial(proto, addr)
+			},
+		},
+		Timeout: time.Second * 5,
+	}
+
+	return &dockerClient{client, network}
+}
+
+func (d *dockerClient) ListContainers() ([]containerInfo, error) {
+
+	const dockerAPIVersion = "v1.41"
+
+	resp, err := d.client.Get(fmt.Sprintf("http://localhost/%s/containers/json", dockerAPIVersion))
```

Comment:
I'm not sure how this work with non-tcp docker, i.e. unix socket

Reply:
I've tested it and http://blah/ will still work as long as it is plain http over some socket. Unfortunately this does not work if user exposes docker socket with http**S** because (1) the scheme needs to be changed to `https` and (2) request hostname must match the one in TLS certificate. My question is do we even care to support this use case? I imagine we can add some config flag indicating that docker should be requested over https.

Reply:
> My question is do we even care to support this use case? I imagine we can add some config flag indicating that docker should be requested over https.

We can postpone it and see if this even a real use case

Reply:
Ok then.